expression/builtin_math_vec.go

```diff
@@ -263,3 +264,37 @@ func (b *builtinRoundDecSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Co
 func (b *builtinRoundDecSig) vectorized() bool {
 	return true
 }
+
+func (b *builtinRoundWithFracDecSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	if err := b.args[0].VecEvalDecimal(b.ctx, input, result); err != nil {
+		return err
+	}
+	buf, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[1].VecEvalInt(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	result.MergeNulls(buf)
+	tmp := new(types.MyDecimal)
+	d64s := result.Decimals()
+	i64s := buf.Int64s()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		if err := d64s[i].Round(tmp, mathutil.Min(int(i64s[i]), b.tp.Decimal), types.ModeHalfEven); err != nil {
```

Comment:
Could we use `&d64s[i]` instead of `tmp` here, to avoiding extra data copy at line 293?
And you can change it at `builtinRoundDecSig` also.

Reply:
I remove tmp like this 
```
	result.MergeNulls(buf)
	d64s := result.Decimals()
	i64s := buf.Int64s()
	for i := 0; i < n; i++ {
		if result.IsNull(i) {
			continue
		}
		if err := d64s[i].Round(&d64s[i], mathutil.Min(int(i64s[i]), b.tp.Decimal), types.ModeHalfEven); err != nil {
			return err
		}
	}
```
but it can't pass unittest, the report is 
```
builtin_math_vec_test.go:66:
    testVectorizedBuiltinFunc(c, vecBuiltinMathCases)
bench_test.go:593:
    c.Assert(*val, Equals, d64s[i])
... obtained types.MyDecimal = types.MyDecimal{digitsInt:9, digitsFrac:0, resultFrac:0, negative:false, wordBuf:[9]int32{18640, 0, 0, 0, 0, 0, 0, 0, 0}}
... expected types.MyDecimal = types.MyDecimal{digitsInt:5, digitsFrac:0, resultFrac:0, negative:false, wordBuf:[9]int32{18640, 0, 0, 0, 0, 0, 0, 0, 0}}
```

Reply:
It seems that the method `MyDecimal.Round` doesn't support this way `self.Round(self, ...)`.

Reply:
https://github.com/pingcap/tidb/blob/8d165f0fd2bddec1d265e05f9abc8c794f219178/types/mydecimal.go#L798

It's documented that it's allowed in that way, could you have another investigation?

Reply:
https://github.com/pingcap/tidb/blob/8d165f0fd2bddec1d265e05f9abc8c794f219178/types/mydecimal.go#L798

It's documented that it's allowed in that way. 
@FateTHarlaown Could you have another investigation?
I think the problem occurs in this line https://github.com/pingcap/tidb/blob/8d165f0fd2bddec1d265e05f9abc8c794f219178/types/mydecimal.go#L834
and in our test framework, we need the two decimals from old and new are deeply equal.

Reply:
could we use `c.Assert(d64s[i].Compare(val), Equals, 0)`to cmpare two Decimals rather than `c.Assert(*val, Equals, d64s[i])`  in test framework ?

Reply:
I'd rather not. But we can add a `TODO` here to reuse `d64s[i]` later.

Reply:
I added a TODO here, is this description reasonable？